Solutions/Marain.Operations.Api.Specs/Steps/OperationsUsingClientSteps.cs

```diff
@@ -0,0 +1,262 @@
+﻿// <copyright file="OperationsUsingClientSteps.cs" company="Endjin Limited">
+// Copyright (c) Endjin Limited. All rights reserved.
+// </copyright>
+
+namespace Marain.Operations.Api.Specs.Steps
+{
+    using System;
+    using System.Net;
+    using System.Threading.Tasks;
+    using Corvus.Testing.SpecFlow;
+    using Marain.Operations.Api.Specs.Bindings;
+    using Marain.Operations.Client.OperationsControl;
+    using Marain.Operations.Client.OperationsControl.Models;
+    using Marain.Operations.Client.OperationsStatus;
+    using Marain.Operations.Client.OperationsStatus.Models;
+    using Marain.TenantManagement.Testing;
+    using Microsoft.Extensions.DependencyInjection;
+    using Microsoft.Rest;
+    using NUnit.Framework;
+    using TechTalk.SpecFlow;
+
+    [Binding]
+    public class OperationsUsingClientSteps : Steps
+    {
+        public const string LastResponseContextKey = "LastResponse";
+
+        [When("I use the operations status client to get the operation with the Id called '(.*)'")]
+        public Task WhenIUseTheOperationsStatusClientToGetTheOperationWithTheIdCalled(string operationIdName)
+        {
+            Guid operationId = this.ScenarioContext.Get<Guid>(operationIdName);
+
+            return this.WhenIUseTheOperationsStatusClientToGetTheOperationWithId(operationId);
+        }
+
+        [When("I use the operations status client to get the operation with Id '(.*)'")]
+        public Task WhenIUseTheOperationsStatusClientToGetTheOperationWithId(Guid id)
+        {
+            IServiceProvider serviceProvider = ContainerBindings.GetServiceProvider(this.FeatureContext);
+            IMarainOperations client = serviceProvider.GetRequiredService<IMarainOperations>();
+            var transientTenantManager = TransientTenantManager.GetInstance(this.FeatureContext);
+
+            return Exceptions.ExecuteAndStoreExceptionAsync(
+                this.ScenarioContext,
+                async () =>
+                {
+                    Client.OperationsStatus.Models.Operation? operation = await client.GetOperationByIdAsync(
+                        transientTenantManager.PrimaryTransientClient.Id,
+                        id).ConfigureAwait(false);
+
+                    this.ScenarioContext.Set(operation, LastResponseContextKey);
+                });
+        }
+
+        [Given("I generate a new operation Id and call it '(.*)'")]
+        public void GivenIGenerateANewOperationIdAndCallIt(string newOperationIdName)
+        {
+            this.ScenarioContext.Set(Guid.NewGuid(), newOperationIdName);
```

Comment:
You know you can just stick all this stuff inside fields in the class, right? You don't have to put all your state in the context.

For a while I think there may have been a bug where that wasn't the case, but you don't need to do it this way now. The one justification is if you need to share stuff across multiple classes, although even then, SpecFlow's lightweight DI-like mechanism is usually simpler.

Reply:
(Obviously you'd use a field with a dictionary to get the exact same effect here, since the mechanism you've used allows multiple operation id names.)

Reply:
Force of habit really - at points in the past, we've had test steps split across more than one class, so sticking values like this in the context was necessary.

Also in this particular example, it seems unnecessary to add an `IDictionary<string, object>` to store values in when this is essentially what the SpecFlow context items are (both ScenarioContext and FeatureContext inherit `Dictionary<string, object>`, but add some generic `Get<T>` and `Set<T>` methods for simplicity.

How strongly do you feel about this?

Reply:
The thing I don't like about using the contexts this way is that its a non-discoverable, vaguely-typed black hole. The advantage of a dedicated dictionary is that it's clear that it's for exactly one thing, as opposed to the "chuck it all in here" context objects.

But I don't feel strongly enough to ask you to change it here. We do this all over the place. I was more hoping to encourage a gradual move away from this practice over time.